File: models/loksta-volterra.py
import numpy as np
import matplotlib.pyplot as plt


n_equations = 2

tol = 1e-6
# Parámetros del modelo
a = 5
b = 0.05
c = 0.0004
d = 0.2


iteration = 0
est = [[], []]


t0 = 0
tf = 50
dt = 0.1
t=[t0]


def deriv(inp):
	x=a*inp[0] - b*inp[0]*inp[1]
	y=c*inp[0]*inp[1] - d*inp[1]

	return [x, y]


def rkf45_step(tt, inp, hh):
    a = [0, 1/4, 3/8, 12/13, 1, 1/2]
    b = [
        [0, 0, 0, 0, 0],
        [1/4, 0, 0, 0, 0],
        [3/32, 9/32, 0, 0, 0],
        [1932/2197, -7200/2197, 7296/2197, 0, 0],
        [439/216, -8, 3680/513, -845/4104, 0],
        [-8/27, 2, -3544/2565, 1859/4104, -11/40]
    ]
    c4 = [25/216, 0, 1408/2565, 2197/4104, -1/5, 0]
    c5 = [16/135, 0, 6656/12825, 28561/56430, -9/50, 2/55]
    
    k = []
    for i in range(6):
        y_temp = [inp[j] + hh * sum(b[i][m] * k[m][j] for m in range(i)) if i > 0 else inp[j] for j in range(n_equations)]
        out=deriv(y_temp)
        k.append(out)
    
    y4 = [inp[j] + hh * sum(c4[i] * k[i][j] for i in range(6)) for j in range(n_equations)]
    y5 = [inp[j] + hh * sum(c5[i] * k[i][j] for i in range(6)) for j in range(n_equations)]
    
    error = np.linalg.norm(np.array(y5) - np.array(y4))
    return y5, error


def simulation():
	global iteration
	est[1].append(90)
	est[0].append(30)
	h = dt
	while t[-1] < tf:
		inp = [est[i][-1] for i in range(n_equations)]
		if t[-1] + h > tf:
			h = tf - t[-1]
		y_new, error = rkf45_step(t[-1], inp, h)
		
		if error <= tol:
			t.append(t[-1] + h)
			for i in range(n_equations):
				est[i].append(y_new[i])
		
		h *= min(5, max(0.2, 0.84 * (tol / error) ** 0.25))

		print(f"Time: {t[-1]:.2f}, Error: {error:.2e}, Step size: {h:.2e}")

if __name__ == '__main__':
	simulation()

	# Guardamos los resultados en un archivo de texto
	with open('results.csv', 'w') as f:
		for i in range(len(t)):
			f.write(f'{t[i]} 	{est[0][i]}	{est[1][i]}	 \n')

File: models/loksta-volterra.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <iomanip>
#include <fstream>
using namespace std;

const int n_equations = 2;

double tol = 1e-6; // Tolerance for RKF45
// Model parameters
const double a = 5;
const double b = 0.05;
const double c = 0.0004;
const double d = 0.2;

const double t0 = 0;
const double tf = 50;
const double dt = 0.1;
vector<double> t = {t0};

vector<vector<double>> est(n_equations);

vector<double> deriv(const vector<double>& inp) {
    vector<double> out(n_equations);
    out[0] = a*inp[0] - b*inp[0]*inp[1];
    out[1] = c*inp[0]*inp[1] - d*inp[1];
    return out;
}

void rkf45_step(double tt, const vector<double>& inp, double hh, vector<double>& y5, double& error) {
    const double a[] = {0, 1.0/4, 3.0/8, 12.0/13, 1.0, 1.0/2};
    const double b[6][5] = {
        {0, 0, 0, 0, 0},
        {1.0/4, 0, 0, 0, 0},
        {3.0/32, 9.0/32, 0, 0, 0},
        {1932.0/2197, -7200.0/2197, 7296.0/2197, 0, 0},
        {439.0/216, -8.0, 3680.0/513, -845.0/4104, 0},
        {-8.0/27, 2.0, -3544.0/2565, 1859.0/4104, -11.0/40}
    };
    const double c4[] = {25.0/216, 0, 1408.0/2565, 2197.0/4104, -1.0/5, 0};
    const double c5[] = {16.0/135, 0, 6656.0/12825, 28561.0/56430, -9.0/50, 2.0/55};

    vector<vector<double>> k(6, vector<double>(n_equations));
    for (int i = 0; i < 6; ++i) {
        vector<double> y_temp(n_equations);
        for (int j = 0; j < n_equations; ++j) {
            y_temp[j] = inp[j];
            for (int m = 0; m < i; ++m) {
                y_temp[j] += hh * b[i][m] * k[m][j];
            }
        }
        k[i] = deriv(y_temp);
    }

    vector<double> y4(n_equations);
    y5.resize(n_equations);
    for (int j = 0; j < n_equations; ++j) {
        y4[j] = inp[j];
        y5[j] = inp[j];
        for (int i = 0; i < 6; ++i) {
            y4[j] += hh * c4[i] * k[i][j];
            y5[j] += hh * c5[i] * k[i][j];
        }
    }

    error = 0.0;
    for (int j = 0; j < n_equations; ++j) {
        error += pow(y5[j] - y4[j], 2);
    }
    error = sqrt(error);
}

int main() {
    for (int i = 0; i < n_equations; ++i) {
        est[i].push_back(0.0);
    }
    est[0][0] = 30;
    est[1][0] = 90;
    double h = dt;
    while (t.back() < tf) {
       vector<double> inp(n_equations);
       for (int i = 0; i < n_equations; ++i) {
           inp[i] = est[i].back();
       }

       if (t.back() + h > tf) {
           h = tf - t.back();
       }
       vector<double> y_new;
       double error;
       rkf45_step(t.back(), inp, h, y_new, error);


       if (error <= tol) {
           t.push_back(t.back() + h);
           for (int i = 0; i < n_equations; ++i) {
               est[i].push_back(y_new[i]);
           }
       }

       h *= min(5.0, max(0.2, 0.84 * pow(tol / error, 0.25)));
       
    cout << scientific << "Tiempo: " << t.back() << " " 
    << "Error: " << error << " Paso: " << h << endl;
    }
    // Save the results to a CSV file
    ofstream results("results.csv");
    results << "Time";
    results << ", x";
    results << ", y";
    results << endl;
    for (size_t i = 0; i < est[0].size(); ++i) {
        results << t[i];
        for (int j = 0; j < n_equations; ++j) {
            results << "	 " << est[j][i];
        }
        results << endl;
    }
    results.close();
    return 0;
}

